thinc/optimizers.py

```diff
@@ -221,28 +238,39 @@ def __call__(
         """
         if len(gradient) < 1:
             return weights, gradient
+
         ops = get_array_ops(weights)
         self.nr_update[key] += 1
         nr_upd = self.nr_update[key]
-        if self.L2 != 0 and not self.L2_is_weight_decay:
-            gradient += self.L2 * weights
-        if self.grad_clip:
-            gradient = ops.clip_gradient(gradient, self.grad_clip)
+        schedule_args = self._schedule_args(key)
+
+        if self.L2(**schedule_args) != 0 and not self.L2_is_weight_decay:
```

Comment:
Could these `(**schedule_args)` parts be handled in `step_schedules`?

Reply:
I wanted to move as much state as possible out of the optimizer. I like that the optimizer only has the number of steps as the changing state, because it allows someone to change the scheduler externally without making it out of sync with the optimizer.

Besides that, we cannot really cache these states in `step_schedules` because they are now parameter-key dependent, so we have to compute them per parameter key.